cli/src/complete.rs

```diff
@@ -0,0 +1,124 @@
+
+use clap::FromArgMatches as _;
+use clap_complete::CompletionCandidate;
+use jj_lib::workspace::DefaultWorkspaceLoaderFactory;
+use jj_lib::workspace::WorkspaceLoaderFactory as _;
+
+use crate::cli_util::expand_args;
+use crate::cli_util::find_workspace_dir;
+use crate::cli_util::GlobalArgs;
+use crate::command_error::internal_error;
+use crate::command_error::CommandError;
+use crate::config::default_config;
+use crate::config::LayeredConfigs;
+use crate::ui::Ui;
+
+pub fn local_bookmarks() -> Vec<CompletionCandidate> {
+    with_jj(|mut jj| {
+        jj.arg("bookmark")
+            .arg("list")
+            .arg("--template")
+            .arg(r#"if(!remote, name ++ "\n")"#)
+            .output()
+            .map(|output| output.stdout)
+            .ok()
+            .and_then(|stdout| String::from_utf8(stdout).ok())
+            .map(|stdout| stdout.lines().map(CompletionCandidate::new).collect())
+            .unwrap_or_default()
+    })
+}
+
+/// Shell out to jj during dynamic completion generation
+///
+/// In case of errors, print them and early return an empty vector.
+fn with_jj<F>(completion_fn: F) -> Vec<CompletionCandidate>
+where
+    F: FnOnce(std::process::Command) -> Vec<CompletionCandidate>,
+{
+    match get_jj_command() {
+        Ok(jj) => completion_fn(jj),
+        Err(e) => {
+            eprintln!("{}", e.error);
+            Vec::new()
+        }
+    }
+}
+
+/// Shell out to jj during dynamic completion generation
+///
+/// This is necessary because dynamic completion code needs to be aware of
+/// global configuration like custom storage backends. Dynamic completion
+/// code via clap_complete doesn't accept arguments, so they cannot be passed
+/// that way. Another solution would've been to use global mutable state, to
+/// give completion code access to custom backends. Shelling out was chosen as
+/// the preferred method, because it's more maintainable and the performance
+/// requirements of completions aren't very high.
+fn get_jj_command() -> Result<std::process::Command, CommandError> {
+    let current_exe = std::env::current_exe().map_err(internal_error)?;
+    let mut command = std::process::Command::new(current_exe);
+
+    // Snapshotting could make completions much slower in some situations
+    // and be undesired by the user.
+    command.arg("--ignore-working-copy");
+
+    // Parse some of the global args we care about for passing along to the
+    // child process. This shouldn't fail, since none of the global args are
+    // required.
+    let app = crate::commands::default_app();
+    let config = config::Config::builder()
+        .add_source(default_config())
+        .build()
+        .map_err(internal_error)?;
+    let mut layered_configs = LayeredConfigs::from_environment(config);
+    let ui = Ui::with_config(&layered_configs.merge())?;
+    let cwd = std::env::current_dir()
+        .and_then(|cwd| cwd.canonicalize())
+        .map_err(internal_error)?;
```

Comment:
nit: `user_error()` seems more appropriate (because it's an environment error.) Applies to other `internal_error()`s.

Reply:
I turned a few internal errors into user errors, but left some that I think the users shouldn'b be able to trigger. Please take another look if it's correct now.